dashboard/src/components/LoginForm/LoginForm.test.tsx

```diff
@@ -0,0 +1,111 @@
+
+const emptyLocation: Location = {
+  hash: "",
+  pathname: "",
+  search: "",
+  state: "",
+};
+
+it("renders a token login form", () => {
+  const wrapper = shallow(
+    <LoginForm authenticated={false} authenticate={jest.fn()} location={emptyLocation} />,
+  );
+  expect(wrapper.find("input#token")).toHaveLength(1);
+  expect(wrapper.find(Redirect)).toHaveLength(0);
+  expect(wrapper).toMatchSnapshot();
+});
+
+it("renders a link to the access control documentation", () => {
+  const wrapper = shallow(
+    <LoginForm authenticated={false} authenticate={jest.fn()} location={emptyLocation} />,
+  );
+  expect(wrapper.find("a").props()).toMatchObject({
+    href: "https://github.com/kubeapps/kubeapps/blob/master/docs/access-control.md",
+    target: "_blank",
+  });
+});
+
+it("updates the token in the state when the input is changed", () => {
+  const wrapper = shallow(
+    <LoginForm authenticated={false} authenticate={jest.fn()} location={emptyLocation} />,
+  );
+  wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
+  expect(wrapper.state("token")).toBe("f00b4r");
+});
+
+describe("redirect if authenticated", () => {
+  it("redirects to / if no current location", () => {
+    const wrapper = shallow(
+      <LoginForm authenticated={true} authenticate={jest.fn()} location={emptyLocation} />,
+    );
+    const redirect = wrapper.find(Redirect);
+    expect(redirect).toHaveLength(1);
+    expect(redirect.props()).toEqual({ push: false, to: { pathname: "/" } });
+  });
+
+  it("redirects to previous location", () => {
+    const location = emptyLocation;
+    location.state = { from: "/test" };
+    const wrapper = shallow(
+      <LoginForm authenticated={true} authenticate={jest.fn()} location={location} />,
+    );
+    const redirect = wrapper.find(Redirect);
+    expect(redirect).toHaveLength(1);
+    expect(redirect.props()).toEqual({ push: false, to: "/test" });
+  });
+});
+
+it("calls the authenticate handler when the form is submitted", () => {
+  const authenticate = jest.fn();
+  const wrapper = shallow(
+    <LoginForm authenticated={false} authenticate={authenticate} location={emptyLocation} />,
+  );
+  wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
+  wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
+  expect(authenticate).toBeCalledWith("f00b4r");
+  expect(wrapper.state("authenticating")).toBe(true);
+});
+
+it("displays an error if the authenticate handler throws an error", async () => {
+  const authenticate = async () => {
+    throw new Error("it's a trap");
+  };
+  const wrapper = shallow(
+    <LoginForm authenticated={false} authenticate={authenticate} location={emptyLocation} />,
+  );
+  wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
+  wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
+
+  // wait for promise to resolve
```

Comment:
@Angelmmiguel I think this was the piece I was missing before, I need to make sure the promise got resolved before verifying the new state.

Reply:
Nice catch! 👍